[Wu Haohui] iP

WuHaohui1231 opened the pull request:
# Jarvis

> “The only way to do great work is to love what you do. If you haven't found it yet, keep looking. Don't settle.” ― Steve Jobs ([Source](https://www.goodreads.com/quotes/772887-the-only-way-to-do-great-work-is-to-love "Source link"))

#### Jarvis is a Personal Assistant Chatbot🤖 that helps a person to keep track of various things.

### Why using Jarvis?
* Text-based
     * Just type in the chatbox with commands⌨️. Jarvis understands you
 * Easy to learn
 * ~~FAST~~ _SUPER_ FAST to use

### All you need to do is,

1. Download it from [here](https://github.com/WuHaohui1231/ip "hey").
2. Double-click it.
3. Type `todo`, `event` or `deadline` to add your tasks.
4. Let it manage your tasks for you 😉 <br/>

And it is __FREE!__ <br/>

### Features:
- [x] View your tasks
- [x] Add a __todo__
- [x] Add a __event__ or __deadline__ with time
- [x] Marking tasks as done/ unmark tasks
- [x] Delete(remove) a task
- [x] Search for tasks by keywords
- [ ] **Reminders** (coming soon)

Prerequisites: JDK 11, update Intellij to the most recent version.

If you Java programmer, you can use it to practice Java too. Here's the main method in the class Launcher:

 ``` java
 public static void main(String[] args) {
     Application.launch(Main.class, args);
 }
 ```

Commit:
Add Gradle support

Commit:
implemented Greet, Echo and Exit function

Commit:
Add, List

Commit:
Mark as Done

Commit:
Level-4

Commit:
Level-5

Commit:
Level-6

Commit:
A-TextUiTesting

Commit:
Level-3 bug fix

WuHaohui1231 committed:
test commit

WuHaohui1231 committed:
Implemented basic task types

WuHaohui1231 committed:
Support task type well

WuHaohui1231 committed:
Add test file

WuHaohui1231 committed:
Handle Error

WuHaohui1231 committed:
Implement delete

WuHaohui1231 committed:
Test branch

WuHaohui1231 committed:
Test Level 8 branch

WuHaohui1231 committed:
Merge branch 'branch-Level-7' into branch-Level-8

WuHaohui1231 committed:
Load and write data on hard disk

Jonaspng commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public void appendLoadedTask(Task task) throws IOException {
+
```
Maybe you can consider removing this empty line?

Jonaspng commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+        DataReadWriter.saveAddedTask(task);
+
+        String msg = "Got it. I've added this task:\n" + "  "
+                + taskList[firstEmptyIndex].toString()
```
Nice indentation 👍

Jonaspng commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+            System.out.println("There is no task with index " + (index + 1));
+            return;
+        }
+        System.out.println("Noted. I've removed this task:\n" +
```
The line break should be placed before the operator.
<img width="725" alt="image" src="https://user-images.githubusercontent.com/60882425/187225273-e978c2de-7e31-4b4c-9a86-9e0866965ee7.png">

Jonaspng reviewed (commented):
Looks good to me. Just some minor issues to fix.

Bacon-Strips commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+    private int firstEmptyIndex = 0;
+
+    public void addTask(String input, Task.TaskType taskType, boolean isDone) throws IOException {
+        Task task = taskType == Task.TaskType.ToDo
+                ? new Todo(input, isDone)
+                : taskType == Task.TaskType.Event
+                ? new Event(input, isDone)
+                : new Deadline(input, isDone);
```
This consecutive ternary operators can be confusing to other coders, might be easier to read if you were to use switch or if else statements directly.

Bacon-Strips commented on src/main/java/DataReadWriter.java:
```diff
@@ -0,0 +1,58 @@
+            }
+            String[] taskInfoArr = taskStr.split("\\|");
+            boolean isDone = taskInfoArr[1].equals("1");
+            if (taskInfoArr[0].equals("T")) {
+                taskList.appendLoadedTask(new Todo(taskInfoArr[2], isDone));
+            } else if (taskInfoArr[0].equals("E")) {
+                taskList.appendLoadedTask(new Event(taskInfoArr[2], taskInfoArr[3], isDone));
+            } else if (taskInfoArr[0].equals("D")) {
+                taskList.appendLoadedTask(new Deadline(taskInfoArr[2], taskInfoArr[3], isDone));
+            } else {
+                throw new RuntimeException("Something wrong with task types in data file");
+            }
```
A switch statement can be used here as well.

Bacon-Strips reviewed (commented):
The code quality seems decent, it looks quite readable

WuHaohui1231 committed:
Add date support

WuHaohui1231 committed:
Add time support

WuHaohui1231 committed:
Merge branch-Level-8

WingKei67 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,96 @@
+                                "e.g. deadline {name} /by {time}");
+                        continue;
+                    }
+                    taskList.addTask(input.substring(9), Task.TaskType.Deadline, false);
```
Would it be better if the nesting could be less deep?

WingKei67 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,96 @@
+            + "|____/ \\__,_|_|\\_\\___|\n";
+
+    static TaskList taskList;
     public static void main(String[] args) {
```
Besides, do you think the method is a bit too long?

WingKei67 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+
+    public void markTaskAsUnDone(int taskNum) {
+        if (taskNum >= firstEmptyIndex) {
+            System.out.println("There is no task with index " + (taskNum + 1));
```
There are some similar cases where you print something directly from this method. Would it be better if you use Ui class to ensure all the statements are at the same level?

WuHaohui1231 committed:
Refactor for OOP

WuHaohui1231 committed:
Change project name

WuHaohui1231 committed:
Wrap classes into package

WuHaohui1231 committed:
Rename to Jarvis

WuHaohui1231 committed:
Start writing JavaDoc

WuHaohui1231 committed:
Merge pull request #3 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support - merge from 2103 course repo

WuHaohui1231 committed:
Configure gradle

WuHaohui1231 committed:
Add JUnit test

WuHaohui1231 committed:
Create JAR

WuHaohui1231 committed:
Partially implement search function

WuHaohui1231 committed:
Use Java ArrayList to store tasks

WuHaohui1231 committed:
Complete search function

WuHaohui1231 committed:
Follow coding standard

WuHaohui1231 committed:
Add JavaDoc. Git stashed change from another branch

WuHaohui1231 committed:
Merge branch-A-CodingStandard and resolve conflict

WuHaohui1231 committed:
Merge branch-Level-9

WuHaohui1231 committed:
Add JavaFX GUI

WuHaohui1231 committed:
Improve GUI

WuHaohui1231 committed:
Level 10 Add GUI

WuHaohui1231 committed:
Use assertions

WuHaohui1231 committed:
Use assertions
No assertions in code previously
To document important assumptions that should hold at various points in the code
Add assetion in data IO and task list processing
To ensure the data file exists before operation and task list is modified correctly

WuHaohui1231 committed:
Merge pull request #4 from WuHaohui1231/branch-A-Assertions

Branch a assertions

WuHaohui1231 committed:
Improve code quality
Have some violation of good code quality before
Refactor code for quality
To help others to understand my code easily
Tag: A-CodeQuality

WuHaohui1231 committed:
Merge pull request #5 from WuHaohui1231/branch-A-CodeQuality

Improve code quality

WuHaohui1231 committed:
Add priority to sort the task list
Previously, the tasks is listed in order of time added
We want to allow users to see 3 groups of tasks, with each group sorted
Let Task implement Comparable and add compareTo method to three kinds of tasks.
So that we can sort the task list when they are changed
This is for the extension C-Priority

WuHaohui1231 committed:
Improve GUI
Add colors. Clip profile image to circle. Beautify chat box
Optimize for window resizing to view more context

WuHaohui1231 committed:
Setup ShadowJar to create JAR with GUI

WuHaohui1231 committed:
Change default picture and upload UI screenshot

WuHaohui1231 committed:
Draft User Guide

WuHaohui1231 committed:
Add help command. Add optional end time for events

WuHaohui1231 committed:
Add help command. Add optional end time for events (modification)

WuHaohui1231 committed:
Add help command. Add optional end time for events. (Correction)

WuHaohui1231 committed:
Add help command. Add optional end time for events. (Correction)

WuHaohui1231 committed:
Use streams

WuHaohui1231 committed:
Update UI screenshot

WuHaohui1231 committed:
Complete User Guide

WuHaohui1231 committed:
Add CheckStyle. Correct style violation

WuHaohui1231 committed:
Update UG

WuHaohui1231 committed:
improve code quality and fix

WuHaohui1231 committed:
Remove dead code

WuHaohui1231 committed:
Merge pull request #7 from WuHaohui1231/branch-Fix-CodeQuality

Fix issues and improve code quality
